### PR TITLE
JsonStringInput: accept decimal format for integer types

### PR DESCRIPTION
**Pull request by ghik**

### CODE REVIEW

File: commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala [hunk 1]
```diff
@@ -43,24 +43,84 @@ class JsonStringInput(
...
+      if (str.charAt(0) == '-') str.length > 1 && loop(1)
+      else loop(0)
+    }
+  }
+
+  private def parseNumber[T](parse: String => T): T = {
+    val str = reader.jsonType match {
+      // accepting JSON strings as numbers (also because they may be Infinity/-Infinity/NaN)
+      case JsonType.number | JsonType.string => reader.currentValue.asInstanceOf[String]
+      case _ => expectedError(JsonType.number)
```
Comment (ddworak) [inline commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala:63]: not tested

File: commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala [hunk 2]
```diff
@@ -43,24 +43,84 @@ class JsonStringInput(
...
+    else {
+      val dbl = str.toDouble
+      if (dbl.isValidByte) dbl.toByte
+      else throw new NumberFormatException(str)
+    }
+  }
+
+  override def readShort(): Short = parseNumber { str =>
+    if (isInteger(str)) str.toShort
+    else {
```
Comment (ddworak) [inline commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala:82]: else branch not tested here and in `readByte`

File: commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala [hunk 3]
```diff
@@ -43,24 +43,84 @@ class JsonStringInput(
...
+
+  def readDouble(): Double =
+    parseNumber(_.toDouble)
+
+  def readBigInt(): BigInt = parseNumber { str =>
+    if (isInteger(str)) BigInt(str)
+    else {
+      val bd = BigDecimal(str, options.mathContext)
+      if (bd.isWhole) bd.toBigInt
+      else throw new NumberFormatException(str)
```
Comment (ddworak) [inline commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala:118]: not covered by tests

File: commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala [hunk 4]
```diff
@@ -95,6 +155,8 @@ class JsonStringInput(
     reader.json.substring(startIdx, endIdx)
   }
 
+  def jsonType: JsonType = reader.jsonType
```
[the remark this answers is not in the document]
Comment (ghik) [reply commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala:158]: I needed that a few times and decided to smuggle it here.

File: commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala [hunk 5]
```diff
@@ -43,24 +43,84 @@ class JsonStringInput(
...
+      val dbl = str.toDouble
+      if (dbl.isValidInt) dbl.toInt
+      else throw new NumberFormatException(str)
+    }
+  }
+
+  def readLong(): Long = parseNumber { str =>
+    if (isInteger(str)) str.toLong
+    else {
+      val bd = BigDecimal(str, options.mathContext)
```
Comment (plokhotnyuk) [inline commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala:101]: Using `BigDecimal` for parsing of `Long` values opens ability for DoS/DoW attacks which exploit O(n^2) complexity of Java's way to parse big numbers from textual representations.

[I've raised an issue for `BigInt` and `BigDecimal` types](https://github.com/AVSystem/scala-commons/issues/88) but now it goes to more widely used types.

Comment (ghik) [reply commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala:101]: @plokhotnyuk do you have a safe `BigInt`/`BigDecimal` parsing implementation that we could lend?

Comment (plokhotnyuk) [reply commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala:101]: [Here](https://github.com/plokhotnyuk/jsoniter-scala/blob/6e3fa06217e18411b85dead678b48c4a2eef8a8e/jsoniter-scala-core/jvm/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonReader.scala#L1585-L1841) is the code which has O(n^1.5) complexity and checks limits for the scale and number of digits. I've used [the following default values](https://github.com/plokhotnyuk/jsoniter-scala/blob/b2fef1806987828eb256ec712e354a546d5e2a00/jsoniter-scala-macros/shared/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala#L59-L63) for them.

**Comment by ghik** [comment]
`BigDecimal` parsing problem will be addressed separately.

**Pull request merged**